cdap-proto/src/main/java/co/cask/cdap/proto/Id.java

```diff
@@ -349,4 +352,84 @@ public int hashCode() {
       return Objects.hashCode(namespace, category, name);
     }
   }
+
+
+  /**
+   * Id.Stream uniquely identifies a stream.
+   */
+  public static final class Stream {
+    private final Namespace namespace;
+    private final String streamName;
+
+    private Stream(final Namespace namespace, final String streamName) {
+      Preconditions.checkNotNull(namespace, "Namespace cannot be null.");
+      Preconditions.checkNotNull(streamName, "Stream name cannot be null.");
+
+      Preconditions.checkArgument(isId(streamName), "Stream name has an incorrect format.");
+
+      this.namespace = namespace;
+      this.streamName = streamName;
+    }
+
+    public Namespace getNamespace() {
+      return namespace;
+    }
+
+    public String getNamespaceId() {
+      return namespace.getId();
+    }
+
+    public String getName() {
+      return streamName;
+    }
+
+    public static Stream from(Namespace id, String application) {
+      return new Stream(id, application);
+    }
+
+    public static Stream from(String namespaceId, String streamName) {
+      return new Stream(Namespace.from(namespaceId), streamName);
+    }
+
+    public String toId() {
+      return String.format("%s.%s", namespace.getId(), streamName);
+    }
+
+    public static Stream fromId(String id) {
+      Iterable<String> comps = Splitter.on('.').omitEmptyStrings().split(id);
+      Preconditions.checkArgument(2 == Iterables.size(comps));
+
+      String namespace = Iterables.get(comps, 0);
+      String streamName = Iterables.get(comps, 1);
+      return from(namespace, streamName);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(namespace, streamName);
```

Comment:
Since this object is used as Key for the map in `ConcurrentStreamWriter`, the `hashCode` and `equals` method needs to be very efficient. Calling `Objects.hashCode` is pretty slow. You better compute the hashCode from the namespace and streamName directly and also cache the value of it.